Add support for private catalog

guangbochen opened the pull request:
Address #13371 
1. support http/s helm catalog server with basic auth.
2. support git clone for private repo with username and password(access token).

Types PR: rancher/types#599
UI Issue: https://github.com/rancher/rancher/issues/16244

StrongMonkey commented:
@vincent99 should we put username and password into a seperate resource like catalogCredential? or is it fine to just put it into catalog api.

StrongMonkey commented on pkg/catalog/manager/catalog_sync.go:
```diff
@@ -18,7 +18,19 @@ func (m *Manager) Sync(key string, obj *v3.Catalog) error {
+	catalog.Spec.RetryCount += 1
+	if catalog.Spec.RetryCount > 5 {
+		return nil
+	}
```
what's the point of setting retryCount? i thought the framework should follow some exponential backoff time...

guangbochen replied:
The catalog controller didn't seem to be stopped by the framework exponential backoff time, not sure it was caused the auto-refresh register, and non-stop reties will causes `too many retires block`.

StrongMonkey replied:
@guangbochen i think we should fix that.....

guangbochen commented:
@StrongMonkey the PR is updated with using `FrogetError` to stop back off.
@vincent99 @alena1108 @cjellick as what daishan mentioned above should we put username and password into a separate resource like catalogCredential?

StrongMonkey commented on pkg/catalog/helm/helm.go:
```diff
@@ -44,7 +67,7 @@ func DownloadIndex(indexURL string) (*RepoIndex, error) {
 	defer resp.Body.Close()
 
 	if resp.StatusCode != http.StatusOK {
```
you should only return forgot error if status code is unauthorized.

guangbochen replied:
updated, thanks.

StrongMonkey commented on pkg/catalog/helm/helm.go:
```diff
@@ -7,10 +7,12 @@ import (
 	"encoding/base64"
 	"encoding/binary"
 	"encoding/hex"
+	"github.com/rancher/norman/controller"
```
this import should be moved to bottom

guangbochen replied:
updated, thanks.

StrongMonkey commented on pkg/catalog/manager/manager.go:
```diff
@@ -100,7 +101,7 @@ func hash(content string) string {
 }
 
-func (m *Manager) prepareGitRepoPath(catalog v3.Catalog) (string, string, error) {
+func (m *Manager) prepareGitRepoPath(catalog v3.Catalog, repoURL string) (string, string, error) {
```
should we also add forgot error logic to git?

guangbochen replied:
![image](https://user-images.githubusercontent.com/4569037/48117777-361ab000-e2a5-11e8-8efe-d30f7f97872e.png)

we may not need to add forgot error logic to the git, if the username and password are incorrect the git ls-remote will return `false` and the logic will will try prepareHelmRepoPath only.

StrongMonkey commented on pkg/catalog/manager/catalog_sync.go:
```diff
@@ -28,9 +29,10 @@ func (m *Manager) Sync(key string, obj *v3.Catalog) error {
 	if commit == catalog.Status.Commit {
 		logrus.Debugf("Catalog %s is already up to date", catalog.Name)
-		if v3.CatalogConditionRefreshed.IsUnknown(catalog) {
+		if !v3.CatalogConditionRefreshed.IsTrue(catalog) {
```
why do you need this change? The idea behind is that we set catalog to unknown when refreshing catalog, and if the commit is same and controller will just set condition to true and skip.

guangbochen replied:
if the commit is same but the credentials(e.g, username and password) are changed and the previous conditional status if false then we also need to update it once as edit catalog CRD will not set condition status to `unknown`.

guangbochen commented:
@StrongMonkey the types PR is in https://github.com/rancher/types/pull/599, and I have addressed some feedbacks in the above, can you please help to review it, thanks.

StrongMonkey commented on pkg/catalog/helm/helm.go:
```diff
@@ -97,11 +125,12 @@ func LoadIndex(repoPath string) (*RepoIndex, error) {
 }
 
-func FetchTgz(url string) ([]v3.File, error) {
+// FetchTgz fetchs artefacts from helm catalog
```
typo: artifacts

StrongMonkey commented:
LGTM

StrongMonkey commented:
also @guangbochen can you squash your commits? once type is merged you should have two commits, one for rancher changes and one for updating vendor.

guangbochen commented:
squshed and updated, thanks.
